### ISSUE TITLE
Increase code coverage

### ISSUE THREAD

**Issue by gregnb** [gregnb/mui-datatables#177]
Hi everyone,

I've been trying to juggle a busy work life while slowly adding new features. As a result, the unit tests have taken a back seat 😢.  If you ever wanted to contribute I would love some help in increasing the code coverage.  

Simply clone the repo and run:

```
npm run test:coverage
```

And pick something that interests you. 

Thank you!

**Comment by DTupalov** [comment]
Hi @gregnb 
After a few critical issues I've tried to solve, I think there is need to some refactor MUIDataTable component to declarative approach instead of imperative. Because we have a lot of features, that are meet in different methods of class component, and visual it hard to detect what should happen in those places.

But it would affect current tests, because there we call internal methods (like `instance.filterUpdate` and etc.). We should test external api, not internal.

I want to refactor MUIDataTables (only logic, not view), but save external API. Do you interest in this work?

**Comment by DTupalov** [comment]
I could give an example after next weekend, I think.

**Comment by gregnb** [comment]
I agree I actually have had to write another table component from scratch for my full time job. In that implementation the table model and its helpers are separated out. I’m really strapped for time but I started toying with that same idea.

Separating out into a “table model” and helpers for the model (sorting, etc) then the MuiDataTable component would call on these to create the model and then handle changes to the model.

I didn’t spent long on it but the model and options are so tightly coupled that I found it to be a little painful. I would love to see some ideas you have? I’m not concerned about the unit tests they are too brittle.

If we had a solid rewrite that separated concerns we could worry about integration level unit tests

**Comment by DTupalov** [comment]
OK, I have idea how it should look like, and would make PR with proof of concept (with basic data and sorting). Then, if you would accept this concept, we could separate features and realize it step by step.

**Comment by gregnb** [comment]
@DTupalov No problem, time is a major killer for me as well. I'll have to revisit this when I get some free time in the new year (hopefully)
